vowpalwabbit/core/src/learner.cc

```diff
@@ -312,77 +326,472 @@ bool ec_is_example_header(const example& ec, label_type_t label_type)
+  THROW(message);
+}
 
-void VW::LEARNER::details::learner_build_diagnostic(VW::string_view this_name, VW::string_view base_name,
-    prediction_type_t in_pred_type, prediction_type_t base_out_pred_type, label_type_t out_label_type,
-    label_type_t base_in_label_type, details::merge_fn merge_fn_ptr, details::merge_with_all_fn merge_with_all_fn_ptr)
+learner* as_singleline(learner* l)
 {
-  if (in_pred_type != base_out_pred_type)
+  if (!l->is_multiline()) { return l; }
+  auto message = fmt::format("Tried to use a multiline reduction as a singleline reduction. Name: {}", l->get_name());
+  THROW(message);
+}
+
+void learner::debug_log_message(polymorphic_ex ex, const std::string& msg)
+{
+  if (ex.is_multiline())
   {
-    const auto message = fmt::format(
-        "Input prediction type: {} of reduction: {} does not match output prediction type: {} of base reduction: {}.",
-        to_string(in_pred_type), this_name, to_string(base_out_pred_type), base_name);
-    THROW(message);
+    VW_DBG(*static_cast<VW::multi_ex&>(ex)[0]) << "[" << _name << "." << msg << "]" << std::endl;
   }
-  if (out_label_type != base_in_label_type)
+  else { VW_DBG(static_cast<VW::example&>(ex)) << "[" << _name << "." << msg << "]" << std::endl; }
+}
+
+// Used as a hook to intercept incorrect calls to the base learner.
+void learner::debug_log_message(const char& /* ec */, const std::string& msg)
+{
+  auto message = fmt::format(
+      "Learner: '{}', function: '{}' was called without first being cast to singleline or multiline.", get_name(), msg);
+  THROW(message);
+}
+
+void learner::learn(polymorphic_ex ec, size_t i)
+{
+  assert(is_multiline() == ec.is_multiline());
+  details::increment_offset(ec, increment, i);
+  debug_log_message(ec, "learn");
+  _learn_f(safe_get_previous_learner(), ec);
+  details::decrement_offset(ec, increment, i);
+}
+
+void learner::predict(polymorphic_ex ec, size_t i)
+{
+  assert(is_multiline() == ec.is_multiline());
+  details::increment_offset(ec, increment, i);
+  debug_log_message(ec, "predict");
+  _predict_f(safe_get_previous_learner(), ec);
+  details::decrement_offset(ec, increment, i);
+}
+
+void learner::multipredict(polymorphic_ex ec, size_t lo, size_t count, polyprediction* pred, bool finalize_predictions)
+{
+  assert(is_multiline() == ec.is_multiline());
+  if (_multipredict_f == nullptr)
   {
-    const auto message =
-        fmt::format("Output label type: {} of reduction: {} does not match input label type: {} of base reduction: {}.",
-            to_string(out_label_type), this_name, to_string(base_in_label_type), base_name);
-    THROW(message);
+    details::increment_offset(ec, increment, lo);
+    debug_log_message(ec, "multipredict");
+    for (size_t c = 0; c < count; c++)
+    {
+      _predict_f(safe_get_previous_learner(), ec);
+      if (finalize_predictions)
+      {
+        pred[c] = std::move(static_cast<VW::example&>(ec).pred);  // TODO: this breaks for complex labels because =
```

Comment:
The only label that doesn't do a correct copy is CCB (which contains a pointer). I think all the predictions should do the right thing. The big problem with that though, is that deep copies are really expensive for these large objects

Reply:
The CCB label implements the copy/move operators so it should work okay